### ISSUE TITLE
Bug: [no-unsafe-enum-comparison ] False positive for flag enums

### ISSUE THREAD

**Issue by JoshMcCullough** [typescript-eslint/typescript-eslint#8177]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.ts&code=KYOwrgtgBAYgTmAlgFygbwFBSgQQA54A2wUAvFAIxQA81UADADRZQBCAhiJ%2B2ZTXRWbYACsHZxeVWlABMzAL4YMAYwD2IAM6p2BYr3hJkAOnxFgAbhXqtUVcgAWwOAZT6EKI6PFLEAMygAFAE6ZlAAZLYOTi7IAJRkpOR2js7ucegsapqqxEaEqgDmAQBEMVCIGlAhucWxlvJAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZonpNYAW1pkOY4nyrIOTdFETRoHaJHBgAviD1A&tsconfig=N4KAviQ&tokens=false

### Repro Code

```TypeScript
enum Fruit {
  Apple = 1 << 0,
  Banana = 1 << 1,
  Pear = 1 << 2,
}

const apple = Fruit.Apple;
const otherFruit = Fruit.Pear

if ((apple & otherFruit) === otherFruit) {
  console.log("Fruit is apple.");
}
```

### ESLint Config

```javascript
{
  "rules": {
    "@typescript-eslint/no-unsafe-enum-comparison": "error"
  }
}
```

### tsconfig

```jsonc
{
}
```

### Expected Result

`@typescript-eslint/no-unsafe-enum-comparison` should not be raised, because the enum types on the left and right are the same.

### Actual Result

`@typescript-eslint/no-unsafe-enum-comparison` is incorrectly raised, so I must turn the rule off when I'd prefer not to.

### Additional Info

Please let me know if I'm missing something here!

**Comment by bradzacher** [comment]
This isn't a false positive - flags enums are actually very loose and unsafe in many ways.

The result of `(apple & otherFruit)` is not `Fruit` - it's `number`. So your comparison is not `Fruit === Fruit` - it's `number === Fruit`. Which is why the rule errors.

This typing makes sense because it's hard for TS to model the bit shifting required to validate that the result of the binary expression is a valid value for the enum. It's also hard for TS to tell that was your intended result as well.

In your example - `apple === 1` and `otherFruit === 4`. So `(apple & otherFruit) === 0`. `0` is not a valid value of `Foo` - hence `number` is the correct type.

Number enums in general are just unsafe because any `number` typed variable may be assigned to a location expecting a number enum. Eg

```ts
enum Foo { A = 1 }
const num: number = 0;
const oops: Foo = num; // oh no

if (x === Foo.A) {}
else {
  // x is typed as never here - even though it's completely reachable code!!!
}
```

In general we strongly recommend against using enums for flag purposes and instead just use `number` and leverage an object instead. That way the unsafety is explicit and you're not presuming things to be safe and exhaustive.

**Comment by JoshMcCullough** [comment]
Thanks for info. I wonder if this could be spelled out more clearly in the docs for the rule? I realize now that it's a complex concept...

> In general we strongly recommend against using enums for flag purposes and instead just use number and leverage an object instead. 

Can you elaborate on that?

**Comment by bradzacher** [comment]
By using an enum it suggests some form of safety and exhaustiveness. For example in the latter sample above - TS believes that the else is unreachable - so it treats the variable as `never`. However we can easily prove that to be untrue and we can prove that the variable could in fact be any `number`.

It's very easy to forget about the unsafety too and trust that TS is correct with its bad assumptions because ofc for string enums it is completely correct.

So what I'd suggest is swapping the enum for an object and removing any illusions of exhaustiveness etc, eg:

```ts
const Fruit = {
  Apple: 1 << 0,
  Banana: 1 << 1,
  Pear: 1 << 2,
};
```

This way your variables are of type `number` and so you won't ever accidentally assume that a variable of type `Fruit` must only be a valid `Fruit` value.
